wdl-analysis/src/handlers/find_all_references.rs

```diff
@@ -0,0 +1,227 @@
+
+    Ok(TargetDefinition {
+        name: token.text().to_string(),
+        kind: parent.kind(),
+        location: definition_location.clone(),
+    })
+}
+
+/// Determines the search scope for finding references based on the target
+/// definition type.
+///
+/// # Returns
+///
+/// A vector of document indices that should be searched for references:
+/// - For local symbols: only the target document
+/// - For imported symbols: the target document + all dependent documents
+/// - For unknown symbol types: defaults to target document only
+fn determine_search_scope(
+    target: &TargetDefinition,
+    graph: &DocumentGraph,
+) -> Vec<petgraph::graph::NodeIndex> {
+    let doc_index = graph
+        .get_index(&target.location.uri)
+        .expect("target document should exist in graph");
+
+    match target.kind {
+        // variables are local to the file.
+        SyntaxKind::BoundDeclNode
+        | SyntaxKind::UnboundDeclNode
+        | SyntaxKind::ScatterStatementNode => {
+            vec![doc_index]
+        }
+        // Things which can be imported from other files
+        SyntaxKind::StructDefinitionNode
+        | SyntaxKind::TaskDefinitionNode
+        | SyntaxKind::WorkflowDefinitionNode => {
+            let mut scope = vec![doc_index];
+            let mut dependents = graph.dependents(doc_index).collect::<Vec<_>>();
+            scope.append(&mut dependents);
+            scope
+        }
+
+        _ => vec![doc_index],
+    }
+}
+
+/// Collects references to the target symbol form a single document.
+fn collect_references_from_document(
+    graph: &DocumentGraph,
+    doc_index: petgraph::graph::NodeIndex,
+    target: &TargetDefinition,
+    encoding: SourcePositionEncoding,
+    locations: &mut Vec<Location>,
+) -> Result<()> {
+    let node = graph.get(doc_index);
+    let document = match node.document() {
+        Some(doc) => doc,
+        None => return Ok(()),
+    };
+
+    let lines = match node.parse_state().lines() {
+        Some(lines) => lines,
+        None => return Ok(()),
+    };
+
+    find_references_in_doc(graph, document, target, lines, encoding, locations)
+}
+
+/// Searches for references to the target symbol within a single document.
+///
+/// 1. Traverse all tokens in the document's CST
+/// 2. Filter for identifier tokens matching the target name
+/// 3. For each match, resolve its definition using goto definition
+/// 4. If the resolved definition matches the target, add the reference location
+fn find_references_in_doc(
+    graph: &DocumentGraph,
+    document: &Document,
+    target: &TargetDefinition,
+    lines: &Arc<LineIndex>,
+    encoding: SourcePositionEncoding,
+    locations: &mut Vec<Location>,
+) -> Result<()> {
+    let root = document.root().inner().clone();
+
+    for token in root
+        .descendants_with_tokens()
+        .filter_map(|el| el.into_token())
+    {
+        if token.kind() == SyntaxKind::Ident && token.text() == target.name {
```

Comment:
Is it _guaranteed_ that variables cannot be shadowed in WDL documents that depend on an upstream document? If so, it might be worth adding a comment here explaining that this invariant must hold. Else, to do it a bit more advanced, you should check for a shadow definition of the same name in this document at a minimum.